p2p/server/server.go

```diff
@@ -147,26 +228,51 @@ func (s *Server) streamHandler(stream network.Stream) {
 	}
 }
 
+func (s *Server) trackClientLatency(start time.Time, err error) {
+	if s.metrics != nil {
+		if err != nil {
+			s.metrics.clientLatencyFailure.Observe(time.Since(start).Seconds())
+		} else {
+			s.metrics.clientLatency.Observe(time.Since(start).Seconds())
+		}
+	}
+}
+
 // Request sends a binary request to the peer. Request is executed in the background, one of the callbacks
 // is guaranteed to be called on success/error.
-func (s *Server) Request(ctx context.Context, pid peer.ID, req []byte, resp func([]byte), failure func(error)) error {
+func (s *Server) Request(
+	ctx context.Context,
+	pid peer.ID,
+	req []byte,
+	resp func([]byte),
+	failure func(error),
+) error {
+	start := time.Now()
 	if len(req) > s.requestLimit {
 		return fmt.Errorf("request length (%d) is longer than limit %d", len(req), s.requestLimit)
 	}
 	if s.h.Network().Connectedness(pid) != network.Connected {
 		return fmt.Errorf("%w: %s", ErrNotConnected, pid)
 	}
 	go func() {
-		start := time.Now()
+		var err error
 		defer func() {
 			s.logger.WithContext(ctx).With().Debug("request execution time",
 				log.String("protocol", s.protocol),
 				log.Duration("duration", time.Since(start)),
+				log.Err(err),
 			)
+			s.trackClientLatency(start, err)
 		}()
```

Comment:
With an errgroup this could be just what comes after the `eg.Wait()`:

```go
eg.Go(func() {
	// most of the code in `go func() { ... }()`
})
err := eg.Wait()
s.logger.WithContext(ctx).With().Debug("request execution time",
	log.String("protocol", s.protocol),
	log.Duration("duration", time.Since(start)),
	log.Err(err),
)
s.trackClientLatency(start, err)
```

Reply:
Alternatively we could even get rid of the `go func()` completely and just make `Request` blocking

Reply:
making request blocking is not a part of this change. note that p2p/server should exist at all, and requests should be made over libp2p streams